### PR TITLE
Fix sftp scp file permission

### PR DESCRIPTION
**Pull request by bagajjal**
https://github.com/PowerShell/Win32-OpenSSH/issues/884

1) Convert the mode properly to file permissions.
If mode has "read" permission then we set the file permission to "read & execute"
If mode has "write" permission then we set the file permission to "Write & Modify"

2) Inherit the file permissions from the parent folder when sftp / scp creates the file on windows.
sftp - put & get.
scp - from local to remote windows machine.

### CODE REVIEW

File: contrib/win32/win32compat/misc.c [hunk 1]
```diff
@@ -1304,3 +1307,68 @@ to_lower_case(char *s)
...
+		PACE_HEADER current_aceHeader = NULL;
+		ACCESS_MASK current_access_mask = 0;
+		int mode_tmp = 0;
+		if (!GetAce(dacl, i, &current_ace)) {
+			debug3("GetAce() failed");
+			goto cleanup;
+		}
+
+		current_aceHeader = (PACE_HEADER)current_ace;
+		/* only interested in Allow ACE */
```
Comment (bingbing8) [inline contrib/win32/win32compat/misc.c:1370]: for this scenario, deny ACE also need to considered. if a perm is denied for user, it should count

File: contrib/win32/win32compat/misc.c [hunk 2]
```diff
@@ -1304,3 +1307,68 @@ to_lower_case(char *s)
...
+		if ((current_access_mask & READ_PERMISSIONS) == READ_PERMISSIONS)
+			mode_tmp |= S_IROTH;
+
+		if ((current_access_mask & WRITE_PERMISSIONS) == WRITE_PERMISSIONS)
+			mode_tmp |= S_IWOTH;
+
+		if ((current_access_mask & EXECUTE_PERMISSIONS) == EXECUTE_PERMISSIONS)
+			mode_tmp |= S_IXOTH;
+
+		if (IsWellKnownSid(current_trustee_sid, WinWorldSid))
```
Comment (bingbing8) [inline contrib/win32/win32compat/misc.c]: I would move this check before retrieving mode. 
if (!IsWellKnownSid(current_trustee_sid, WinWorldSid) && !IsWellKnownSid(current_trustee_sid, WinAuthenticatedUserSid))
continue;

**Review by bingbing8 — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/misc.c [hunk 3]
```diff
@@ -1308,15 +1311,36 @@ to_lower_case(char *s)
...
+get_final_mode(int allow_mode, int deny_mode)
+{	
+	// If deny permissions are not specified then return allow permissions.
+	if (!deny_mode) return allow_mode;
+
+	// If allow permissions are not specified then return allow permissions (0).
+	if (!allow_mode) return allow_mode;
+	
+	if(deny_mode & S_IROTH)
+		allow_mode = (allow_mode | S_IROTH) ^ S_IROTH;
```
Comment (bingbing8) [inline contrib/win32/win32compat/misc.c:1324]: How this work? 
does this work? allow_mode & ~S_IROTH

File: contrib/win32/win32compat/misc.c [hunk 4]
```diff
@@ -1344,29 +1368,44 @@ get_others_file_permissions(wchar_t * file_name)
 
 		current_aceHeader = (PACE_HEADER)current_ace;
 		/* only interested in Allow ACE */
-		if (current_aceHeader->AceType != ACCESS_ALLOWED_ACE_TYPE)
+		if (!(current_aceHeader->AceType == ACCESS_ALLOWED_ACE_TYPE || 
+		    current_aceHeader->AceType == ACCESS_DENIED_ACE_TYPE))
 			continue;
 
 		PACCESS_ALLOWED_ACE pAllowedAce = (PACCESS_ALLOWED_ACE)current_ace;
```
Comment (bingbing8) [inline contrib/win32/win32compat/misc.c]: if current_aceHeader is of Denied_ACE_Type, does this statement "PACCESS_ALLOWED_ACE pAllowedAce = (PACCESS_ALLOWED_ACE)current_ace" success?
Suggest to this line:
if(current_aceHeader->AceType == ACCESS_ALLOWED_ACE_TYPE)
{
    pAllowedAce = (PACCESS_ALLOWED_ACE)current_ace;
   current_trustee_sid = &(pAllowedAce->SidStart);
}
else if(current_aceHeader->AceType == ACCESS_DENIED_ACE_TYPE)
{
     pDeniedAce = (PACCESS_DENIEDED_ACE)current_ace;
    current_trustee_sid = &(pDeniedAce->SidStart);
}

**Review by bingbing8 — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/misc.c (hunk 4 above)
Comment (bagajjal) [reply contrib/win32/win32compat/misc.c]: PACCESS_ALLOWED_ACE, PACCESS_DENIED_ACE are same.. I have tested this.. Just to minimize the code, I didn't have this if/else loop. if you want to be more readable then I can add.

here is the definition.

typedef struct _ACCESS_ALLOWED_ACE {
    ACE_HEADER Header;
    ACCESS_MASK Mask;
    DWORD SidStart;
} ACCESS_ALLOWED_ACE;

typedef ACCESS_ALLOWED_ACE *PACCESS_ALLOWED_ACE;

typedef struct _ACCESS_DENIED_ACE {
    ACE_HEADER Header;
    ACCESS_MASK Mask;
    DWORD SidStart;
} ACCESS_DENIED_ACE;
typedef ACCESS_DENIED_ACE *PACCESS_DENIED_ACE;

Comment (manojampalam) [reply contrib/win32/win32compat/misc.c]: Add a comment about PACCESS_ALLOWED_ACE, PACCESS_DENIED_ACE  being the same for code clarify. Merging this now, so please add it in a subsequent PR.

**Pull request merged**